### PR TITLE
Perf/Reduce RPC deserialization overhead

### PR DESCRIPTION
**Pull request by asdacap**
- Slightly reduce deserialization overhead by not serializing params back to string, instead storing the JToken.
- Benchmark shows avg time reduced from 7.6ms to 5.95ms and reduced heap allocated from 2 GB to 1.67 GB (I don't know how its measured).

## Changes:
- Add benchmark
- Store params as JToken instead of string.

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [X] Perf
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [X] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [ ] Yes
- [X] No

**Comments about testing , should you have some** (optional)

- Unit tests passed. 

## Further comments (optional)

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: src/Nethermind/Nethermind.JsonRpc/JsonRpcProcessor.cs [hunk 1]
```diff
@@ -91,46 +91,67 @@ private IEnumerable<(JsonRpcRequest Model, List<JsonRpcRequest> Collection)> Des
...
+            if (idField != null)
             {
-                paramsToken = token.SelectToken("Params");
-                if (paramsToken is null)
+                // Not sure what is the logic here.. not gonna unuse the IdConverter
+                JTokenReader reader = new JTokenReader(idField);
+                if (reader.Read())
                 {
-                    return;
+                    jsonRpcRequest.Id = new IdConverter().ReadJson(reader, null, null, null);
```
Comment (flcl42) [inline src/Nethermind/Nethermind.JsonRpc/JsonRpcProcessor.cs:132]: `new IdConverter()` can be cached.

File: src/Nethermind/Nethermind.JsonRpc/JsonRpcProcessor.cs [hunk 2]
```diff
@@ -91,46 +91,67 @@ private IEnumerable<(JsonRpcRequest Model, List<JsonRpcRequest> Collection)> Des
...
+        /// Manually convert JToken to JsonRpcRequests as the param inside JsonRpcRequests is a JToken. If we use
+        /// a JTokenReader, it would iterate through the JSON which is a waste of time, as we already have a JToken.
+        /// </summary>
+        /// <param name="token"></param>
+        /// <returns></returns>
+        private JsonRpcRequest JTokenToJsonRpcRequest(JToken token)
         {
-            var paramsToken = token.SelectToken("params");
-            if (paramsToken is null)
+            JsonRpcRequest jsonRpcRequest = new JsonRpcRequest();
```
Comment (flcl42) [inline src/Nethermind/Nethermind.JsonRpc/JsonRpcProcessor.cs:111]: <sub>Can be just `new();` and below one more</sub>

File: src/Nethermind/Nethermind.JsonRpc/JsonRpcProcessor.cs [hunk 3]
```diff
@@ -91,46 +91,67 @@ private IEnumerable<(JsonRpcRequest Model, List<JsonRpcRequest> Collection)> Des
...
+                JTokenReader reader = new JTokenReader(idField);
+                if (reader.Read())
                 {
-                    return;
+                    jsonRpcRequest.Id = new IdConverter().ReadJson(reader, null, null, null);
                 }
             }
 
-            if (paramsToken is JValue)
+            JToken? paramsField = token["params"];
```
Comment (flcl42) [inline src/Nethermind/Nethermind.JsonRpc/JsonRpcProcessor.cs:136]: Should we check for "Jsonrpc"/"JsonRpc"/"Id"/etc too?
Btw, can be `token["params"] ?? token["params"]`

**Review by flcl42 — CHANGES_REQUESTED** [review]

File: src/Nethermind/Nethermind.JsonRpc/JsonRpcProcessor.cs (hunk 3 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.JsonRpc/JsonRpcProcessor.cs:136]: I'm confused as to why we check the capitalized case in the first place. The spec mentioned them all as small case. https://www.jsonrpc.org/specification. Should we check regardless?

File: src/Nethermind/Nethermind.JsonRpc/JsonRpcProcessor.cs (hunk 2 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.JsonRpc/JsonRpcProcessor.cs]: Done

File: src/Nethermind/Nethermind.JsonRpc/JsonRpcProcessor.cs (hunk 1 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.JsonRpc/JsonRpcProcessor.cs]: Done

File: src/Nethermind/Nethermind.JsonRpc/JsonRpcProcessor.cs (hunk 3 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.JsonRpc/JsonRpcProcessor.cs]: I'll just check it regardless.

**Review by flcl42 — APPROVED** [review]

**Comment by asdacap** [comment]
I have another idea. Will try that first.

**Comment by asdacap** [comment]
Made some mistake with the benchmark. Used limbologs which trigger debug output which trigger deserialization again, and used some nsubstiture which I suspect have some affect on performance.

Latest result, with CPU governor explicitly set to performance shows that this reduce the overhead by half and memory allocation by about 30%. But then again, its only 1ms per call... so.... its something. 

Another approach is to replace the param string with a pooled `MemoryStream`. This perform slightly better, but uses much more complicated code.

**Pull request merged**